cmd/zoekt-sourcegraph-indexserver/fetcher.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/google/zoekt"
+	"io"
+	"os/exec"
+	"strings"
+)
+
+// a deltaFetcher returns a list of changed files along with their file contents
+//
+// NOTE: The idea of returning the file contents in the same function seems awkward
+// with local checkouts, but necessary in a remote one
+type deltaFetcher interface {
+	// Fetch returns the paths that have changed on the latest
+	// commit for each branch, along with their file contents.
+	Fetch(ctx context.Context, branches ...zoekt.RepositoryBranch) (branchDeltas, error)
+}
+
+type branchDeltas map[string]delta
+
+// delta contains all the paths that have been added, changed, and removed
+// along with the file contents
+//
+// TODO@ggilmore: should the file contents  be io.ReadClosers, and should delta offer a "close"
+// method that closes all the file contents (thinking about the local file case)
+type delta struct {
+	// Added is a list of all the paths that have been added,
+	// along with their file contents.
+	Added map[string]io.Reader
+
+	// Changed is a list of all the paths that have been changed,
+	// along with their file contents.
+	Changed map[string]io.Reader
+
+	// Removed is a list of all the paths that were removed.
+	Removed []string
+}
+
+// NOTE: We can mimic the existing behavior (albeit with extra work on Zoekt's end)
+// by creating a fresh git checkout every time, feeding it to local fetcher,
+// and copying all the files into a new repository for zoekt-git-index to process. This
+// doesn't seem that useful, but we can do this to be able to merge a PR earlier.
+//
+// The delta will appear as if all files were deleted and then re-added on every commit.
+//
+// We can have another service whose job is just to manage a permanent local checkout (deal with corruption,
+// garbage collection, evicting epos, etc.) but the implementation of localFetcher doesn't
+// care about where this service comes from.
```

Comment:
if we go the route of having a local repo, we should probably more directly change `zoekt-git-index` since it currently directly interacts with git via `go-git`. IE it is the cleaner place for this sort of logic.

There is even the idea of having a `go-git` remote which is gitserver. This might be quite a useful thing to implement, because then it is all quite pure and the code is pretty much the same for local vs remote. Would you be interested in exploring this / pairing on this idea? It may be a very powerful abstraction to use in other places as well. cc @stefanhengl who also liked this idea.